Examples/Io/HepMC3/src/HepMC3Reader.cpp

```diff
@@ -17,3 +20,47 @@ bool ActsExamples::HepMC3ReaderAscii::readEvent(
 bool ActsExamples::HepMC3ReaderAscii::status(HepMC3::ReaderAscii& reader) {
   return !reader.failed();
 }
+
+ActsExamples::HepMC3ReaderAscii::HepMC3ReaderAscii(
+    const ActsExamples::HepMC3ReaderAscii::Config& cfg,
+    Acts::Logging::Level lvl)
+    : m_cfg(cfg),
+      m_eventsRange(determineEventFilesRange(
+          cfg.inputDir, cfg.inputStemFileName + ".hepmc3")),
+      m_logger(Acts::getDefaultLogger("HepMC3ReaderAscii", lvl)) {
+  if (m_cfg.inputStemFileName.empty()) {
+    throw std::invalid_argument("Missing input filename stem");
+  }
+  if (m_cfg.outputCollection.empty()) {
+    throw std::invalid_argument("Missing output collection");
+  }
+}
+
+std::string ActsExamples::HepMC3ReaderAscii::HepMC3ReaderAscii::name() const {
+  return "HepMC3ReaderAscii";
+}
+
+std::pair<size_t, size_t> ActsExamples::HepMC3ReaderAscii::availableEvents()
+    const {
+  return m_eventsRange;
+}
+
+ActsExamples::ProcessCode ActsExamples::HepMC3ReaderAscii::read(
+    const ActsExamples::AlgorithmContext& ctx) {
+  std::vector<std::shared_ptr<HepMC3::GenEvent>> events;
+  HepMC3::GenEvent event(HepMC3::Units::GEV, HepMC3::Units::MM);
+
+  auto path = perEventFilepath(m_cfg.inputDir, m_cfg.inputStemFileName + ".csv",
+                               ctx.eventNumber);
+  HepMC3::ReaderAscii reader(path);
+
+  while (reader.read_event(event)) {
+    events.push_back(std::make_shared<HepMC3::GenEvent>(event));
+    event.clear();
+  }
+
+  ctx.eventStore.add(m_cfg.outputCollection, std::move(events));
+
+  return reader.failed() ? ActsExamples::ProcessCode::ABORT
+                         : ActsExamples::ProcessCode::SUCCESS;
```

Comment:
If the reading has a failure the output collection might be missing something. The failure check should therefore come before the collection is written to the whiteboard.

Reply:
fixed